uint256: 1 is a constant

ajtowns opened the pull request:
`UINT256_ONE()` returns a reference to a global; mark it as const to be sure someone doesn't accidently modify it.

ajtowns committed:
uint256: Update constructors to c++11, make ONE static

Replace the memset with C++11 value/aggregate initialisation of
the m_data array, which still ensures the unspecified values end
up as zero-initialised.

This then allows changing UINT256_ONE() from dynamically allocating an
object, to a simpler referencing a static allocation.

ajtowns commented:
This was added in #17261 in which it was [remarked](https://github.com/bitcoin/bitcoin/pull/17261#discussion_r363058528) "static initialization order result[ed] in a potential uninitialized object access." I have an alternative change at https://github.com/ajtowns/bitcoin/commits/202009-uint256_1_v2 which does a static init and hence avoids the functional call, and seems to pass travis fine, so might be worth considering that? cc @achow101 @kallewoof

achow101 commented:
> I have an alternative change at https://github.com/ajtowns/bitcoin/commits/202009-uint256_1_v2 which does a static init and hence avoids the functional call, and seems to pass travis fine, so might be worth considering that?

Concept ACK

ajtowns committed:
scripted-diff: Replace UINT256_ONE() with uint256::ONE

-BEGIN VERIFY SCRIPT-
sed -i '/inline.* UINT256_ONE() {/,+1d' src/uint256.h
sed -i 's/UINT256_ONE()/uint256::ONE/' $(git grep -l UINT256_ONE)
-END VERIFY SCRIPT-

ajtowns committed:
wallet: no need for duplicate storage for ABANDON_HASH constant

kallewoof commented:
Clear improvement.

ACK 832109af0d213239b1b803a1de7f78f193d28ad2

ajtowns commented:
Updated to simplfy constructors and replace `UINT256_ONE()` call with `uint256::ONE` constant

practicalswift commented:
Concept ACK

kristapsk reviewed (approved):
ACK 915c1c25fd6d6bdc249b71be12db0744312729da

sipa commented:
utACK 915c1c25fd6d6bdc249b71be12db0744312729da

> 1 is a constant

It is in tautologies that the ordering of society is centered.

MarcoFalke commented on src/uint256.h:
```diff
@@ -20,10 +20,11 @@ class base_blob
+    base_blob() : m_data() {}
+
+    /* constructor for constants between 1 and 255 */
+    explicit constexpr base_blob(uint8_t v) : m_data{v} {}
```
```suggestion
    constexpr explicit base_blob(uint8_t v) : m_data{v} {}
```

style nit: Would be nice to use the same ordering of keywords for different lines in the same patch. 

<!--
(https://en.cppreference.com/w/cpp/utility/optional/optional)

ajtowns replied:
Makes sense; https://github.com/cplusplus/draft/wiki/Specification-Style-Guidelines#formatting-declarations-and-definitions also suggests `constexpr explicit`

maflcko reviewed (approved):
Not sure how this could have led to memory access violations. Though, if the wallet or the script interpreter is running after the end of the main function, something else has gone wrong and should probably be fixed.


cr ACK 915c1c25fd6d6bdc249b71be12db0744312729da 📺

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

cr ACK 915c1c25fd6d6bdc249b71be12db0744312729da 📺
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhQlQwAqTJdOujLC6CD+i1XI8ISvOTG76CLhzgK5wg+63FYUVILodO1T0ig7NJH
D+RHnv0fdyO3OIFGH3C4ooSkHNqIiKec7Q0V2uemtU5TZX04jM6ocsP/frsCfSmU
p13UkuNxx5QSnrSlFLxrGrmuX/BkLxdubGLTXNQHNltx5rHz2p0H03ueep2/XrXM
BeMAdrBcBvjsQ3SW5XuUND8hza6YajWpAMOUSnS7rbT/hpxi3SakjYyA8E/fMQUo
tr7F/XVLX9KTYzRFvbQLRdUs2Uy3jIP4AhuJJqa7rt/1wv3Au7N0nL96wPJlBiKO
radqu65JIp2sNDl4XtXcq1gol+O184rASiaeDZlI1kAxUKoBvDKd9XtBnDWHywyl
vvulJWWkZLHN6q0z3JJ2GON8Uyr+q6ErUrOuOyLsuRvnrmawqkc6tBaUpkCefPuQ
6kKwWybrM5xSq0/OYiOQ0vFRJY34GLy7qNmyUOIQrGpbfn9AL7c9B5p9FueqMukP
/HlO1n43
=ZuQH
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `ed0494b57cec88e0ff8da75d55c543250f502605792f0e336b3c5d14e6a470b4  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108ed0494b57cec88e0ff8da75d55c543250f502605792f0e336b3c5d14e6a470b4f010d302f71a536538092dc7d19e7b9e6e2108fff010524f8ba3e85da7edc29cadac69c04e8608f1045f6fa371f0080002ddf0bd286bf10083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010e06cb8053812a419c81bc67f1de48a9508f1045f6fa371f008d55d610f39a5fb040083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010eb1fdf023155bfe63acd1dcf8326977108f1045f6fa371f0088c21a674d65ef3480083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010f6768ac9cf10286e3bdcb774ecf7f60708f1045f6fa371f008d8c14e324cf068c10083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented:
(1 is not a constant in python, but we don't use that here https://kate.io/blog/2017/08/22/weird-python-integers/)

promag reviewed (commented):
ACK 915c1c25fd6d6bdc249b71be12db0744312729da.

hebasto reviewed (approved):
ACK 915c1c25fd6d6bdc249b71be12db0744312729da, I have reviewed the code and it looks OK, I agree it can be merged.

May I suggest a further step:
```diff
--- a/src/uint256.h
+++ b/src/uint256.h
@@ -21,7 +21,7 @@ protected:
     uint8_t m_data[WIDTH];
 public:
     /* construct 0 value by default */
-    base_blob() : m_data() {}
+    constexpr base_blob() : m_data() {}
 
     /* constructor for constants between 1 and 255 */
     explicit constexpr base_blob(uint8_t v) : m_data{v} {}
@@ -112,7 +112,7 @@ public:
  */
 class uint160 : public base_blob<160> {
 public:
-    uint160() {}
+    constexpr uint160() {}
     explicit uint160(const std::vector<unsigned char>& vch) : base_blob<160>(vch) {}
 };
 
@@ -123,7 +123,7 @@ public:
  */
 class uint256 : public base_blob<256> {
 public:
-    uint256() {}
+    constexpr uint256() {}
     constexpr explicit uint256(uint8_t v) : base_blob<256>(v) {}
     explicit uint256(const std::vector<unsigned char>& vch) : base_blob<256>(vch) {}
     static const uint256 ONE;
```
?

ajtowns commented on src/test/uint256_tests.cpp:
```diff
@@ -278,4 +278,10 @@ BOOST_AUTO_TEST_CASE( operator_with_self )
     BOOST_CHECK(v == UintToArith256(uint256S("0")));
 }
 
+BOOST_AUTO_TEST_CASE( check_ONE )
```
Added the `constexpr` annotations. Left the formatting consistent with the rest of uint256_tests.cpp (and arith_uint256_tests.cpp).

kallewoof reviewed (commented):
ACK 4cc7171c9887e88027642927c979e507d7b78dda

promag commented:
ACK 4cc7171c9887e88027642927c979e507d7b78dda

MarcoFalke commented:
re ACK 4cc7171c98, only change is some constexpr shuffling 🛁

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re ACK 4cc7171c98, only change is some constexpr shuffling 🛁
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhxqwwAxiSleGr6i0+dxWiXOWdI69VYKTpCJZx1eqemwQOuNA+YguYEgdIvm+78
2uFRqv3WgLDF7RvJuGVoX8x3BYMaNML0SsADXDNi0A/1FCHMdH1Mh9R+kJLFrYyK
+EsRdPGYnt4RNstY/lKkMiN6zKwEA889hcUFJbGHdnjO9+LsnfO2oCV2HSB5n6c+
uj//8F3NwvyIFwVVYdn0XnDML777HOY6qwsIQFtjNV8MCzPyEdO9qGAyrORWKLIs
rDNUFByTi+i8Eg0iOUeDHarOq9JRRXk9iw09Mo207PeplnjoIvFUbdDLBkpiaIf2
6xh2tIsnZRBcVWyLmwtNPA+4Hg59a6mujwV0HEgoq1gZ55fS6qxB9gHyVElWmqZX
beGvhIM6Oa4j6449UFbGhAMn5hl7m3xXZyGr+HuV8u1RYEpMEpkiNV3CMrMSPQbA
5LY4cIsOLWmuuI/RQJmecftKiKu+YkoknRNjqF7TIqdX5qVMP6V4buB6MeFvB6R/
vgDgxnQp
=JHzT
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `7017836e4e8e4ce0003284c993c86597e80a695b740243eca5ab7c60803d2d5e  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401087017836e4e8e4ce0003284c993c86597e80a695b740243eca5ab7c60803d2d5ef010441e514642a23f8178ad18781a1518bd08fff0100903d8dececb7d17e83deb896503d8c408f1045f718d9cf008dc2f7f82cbbded520083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01038d0b6694b5c8938d1cf5808f62478e508f1045f718d9df008c480ce757d3621200083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010d3da79d3bdbbdf5e0567c1c4f2816aab08f1045f718d9cf00837f2ad926a9516060083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010d79f1e7e9b7900804121a9ffc0ea532908f02065049590e6883717642e3ee945045d6952659e77278b60846bad71cefa6897f308f020e4b41893878b421738daa13da427af6208c5c829f955e285c6b7fb9ede96f92208f1045f718d9df008e572568dc81298f50083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

hebasto reviewed (commented):
re-ACK 4cc7171c9887e88027642927c979e507d7b78dda, only suggested changes since my [previous](https://github.com/bitcoin/bitcoin/pull/20016#pullrequestreview-497059064) review.
